File: apiutils/src/main/java/com/sky/http/UseCase.java
package com.sky.http;

import com.sky.api.OnRequestCallback;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

/**
 * Created by SKY on 2017/5/31.
 */
public abstract class UseCase<T> {
    //    protected Observable<T> observable;
    protected abstract Observable<T> buildObservable();

    public void subscribe(OnRequestCallback<T> request) {
        buildObservable()
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new DefaultSubScriber(request));
    }

    public void unSubscirbe() {
//        observable.unsubscribeOn()
//        observable.unsubscribeOn()
    }


}

File: app/src/main/java/com/sky/oa/utils/http/UseCase.kt
package com.sky.oa.utils.http

import com.sky.oa.api.OnRequestCallback

import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

/**
 * Created by SKY on 2017/5/31.
 */
abstract class UseCase<T : com.sky.sdk.net.http.ApiResponse<*>> {
    protected abstract fun buildObservable(): Observable<T>

    fun subscribe(request: OnRequestCallback<T>) {
        buildObservable()
                .subscribeOn(Schedulers.io())
                .unsubscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(DefaultSubScriber(request))
    }
}
